src/spatio_temporal_voxel_layer.cpp

```diff
@@ -716,10 +723,12 @@ void SpatioTemporalVoxelLayer::updateBounds( \
   ObservationsResetAfterReading();
   current_ = current;
 
+  std::unordered_set<volume_grid::occupany_cell> cleared_cells;
```

Comment:
Why an `unordered_set`? Why not a vector for instance? Was this experimentally shown to be faster?

Reply:
I chose `unordered_set` to avoid duplicates. When clearing the world points, we don't care about the Z coordinate. If we clear world point (1, 2, 3) and (1, 2, 4), they're both gonna be in the same costmap cell, so there is no need to store it and iterate through it twice.

Reply:
OK, that  seems reasonable. I'd ask though if you'd test both and use which ever is practically faster. The STL Vector implementation is very cache friendly, so even areas where you wouldn't think it would be faster to use (because of duplications for instance), it ends up being faster due to continuous memory buffers (I ran into that alot while developing this: what you "think" should be faster based on theory and what is actually faster are not always intuitive). I don't know if it would be faster or not, so please test and let me know. 

You can create an update timer pretty easily with chrono 
```
StartTime = std::chrono::system_clock::now();
... do work
endTime = std::chrono::system_clock::now();
std::cout << "time took " std::chrono::duration_cast<std::chrono::milliseconds>(endTime - StartTime).count() << " ms."<< std::endl;
```

Reply:
You're right, `vector` is **a lot** faster than `unordered_set` when iterating through its contents, even though its size is a lot bigger due to duplicated elements. However, `vector` is only faster if we don't do anything while iterating. Since we need to touch every element, then `unordered_set` becomes the faster alternative since it has fewer elements - meaning less calls to `touch`.

I recorded two small clips where you can see the time it takes for both cases, when decay time is set to 0 (i.e. instant clearing):

* [looping only](https://streamable.com/3mi5zd)
* [looping and touching every element](https://streamable.com/6ysvm6)

Reply:
What was the looping testing? Just the loop without `touch` (so empty `for`?) 

Keep in mind its not just access for iteration, its also about insertion time since vector insertion would probably be faster due to no hash table lookups for existing keys. You should try to measure the full update and take some metrics based on that since the insertions are pretty built into the clearing function

Reply:
I averaged 450 update cycles with the following scenario:
![image](https://user-images.githubusercontent.com/15384781/115018704-18a79280-9eb0-11eb-9896-267bf7f75cdf.png)

The start time was set right before this `if` statement: https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_layer.cpp#L720-L723

and end time at the very bottom of the function, right before `return`.

If I set the start time at the top of the function the execution times were **very** inconsistent, I'm guessing due to the calls to `GetObservations` methods. I don't think my changes should affect the performance before we clear anyway, so I set the start time after the observations are obtained, in order to get more consistent times.

Results:

* **baseline:** 6483 microseconds
* **unordered_set:** 7613 microseconds
* **vector:** 7779 microseconds

So `unordered_set` and `vector` have basically the same performance, which was ~1ms worse than the baseline.